source/server/admin/stats_render.cc

```diff
@@ -262,5 +272,163 @@ void StatsJsonRender::collectBuckets(const std::string& name,
+  // https://prometheus.io/docs/concepts/data_model/
+  return absl::StrCat("envoy_", sanitizeName(extracted_name));
+}
+
+std::string PrometheusStatsRender::sanitizeName(const absl::string_view name) {
+  // The name must match the regex [a-zA-Z_][a-zA-Z0-9_]* as required by
+  // prometheus. Refer to https://prometheus.io/docs/concepts/data_model/.
+  // The initial [a-zA-Z_] constraint is always satisfied by the namespace prefix.
+  return prometheusRegex().replaceAll(name, "_");
+}
+
+std::string PrometheusStatsRender::sanitizeValue(const absl::string_view value) {
+  // Removes problematic characters from Prometheus tag values to prevent
+  // text serialization issues. This matches the prometheus text formatting code:
+  // https://github.com/prometheus/common/blob/88f1636b699ae4fb949d292ffb904c205bf542c9/expfmt/text_create.go#L419-L420.
+  // The goal is to replace '\' with "\\", newline with "\n", and '"' with "\"".
+  return absl::StrReplaceAll(value, {
+                                        {R"(\)", R"(\\)"},
+                                        {"\n", R"(\n)"},
+                                        {R"(")", R"(\")"},
+                                    });
+}
+
+template <class StatType>
+void PrometheusStatsRender::outputStatType(
+    Buffer::Instance& response, const std::vector<StatType>& metrics,
+    const std::string& prefixed_tag_extracted_name,
+    const std::function<std::string(
+        const StatType& metric, const std::string& prefixed_tag_extracted_name)>& generate_output,
+    absl::string_view type) {
+  response.add(fmt::format("# TYPE {0} {1}\n", prefixed_tag_extracted_name, type));
+  for (const auto& metric : metrics) {
+    response.add(generate_output(metric, prefixed_tag_extracted_name));
+  }
+}
+
+template <class StatType>
+std::string
+PrometheusStatsRender::generateNumericOutput(const StatType& metric,
+                                             const std::string& prefixed_tag_extracted_name) {
+  const std::string tags = formattedTags(metric->tags());
+  return absl::StrCat(prefixed_tag_extracted_name, "{", tags, "} ", metric->value(), "\n");
+}
+
+std::string
+PrometheusStatsRender::generateTextReadoutOutput(const Stats::TextReadoutSharedPtr& metric,
+                                                 const std::string& prefixed_tag_extracted_name) {
+  auto tags = metric->tags();
+  tags.push_back(Stats::Tag{"text_value", metric->value()});
+  const std::string formTags = formattedTags(tags);
+  return absl::StrCat(prefixed_tag_extracted_name, "{", formTags, "} 0\n");
+}
+
+std::string
+PrometheusStatsRender::generateHistogramOutput(const Stats::HistogramSharedPtr& metric,
+                                               const std::string& prefixed_tag_extracted_name) {
+  auto parent_histogram = dynamic_cast<Stats::ParentHistogram*>(metric.get());
+  if (parent_histogram != nullptr) {
+    // TODO(rulex123): support for the 3 histogram "bucket modes" offered in other
```

Comment:
This makes sense.

FYI in https://github.com/envoyproxy/envoy/pull/26472 I have added a new bucket mode tentatively called 'detailed' which will have the raw bucket data. That *may* also be useful for Prom as you can see a lot more interesting data with this IMO, but it might complicate aggregation in Prom across multiple Envoys.

Don't worry Greg -- that PR is still a ways away from being atomized into probably around 4 for review, but I think the first PR will be the addition of a new bucket-mode to show the original buckets computed by circlhist.

Nevertheless I agree removing this TODO is fine for now.